### PR TITLE
Tripwire

### PR DESCRIPTION
**Pull request by orhovy**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/13722

## Description
Tripwire new integration of FIM product.
Contains 4 commands: 
- tripwire-nodes-list
- tripwire-rules-list
- tripwire-elements-list
- tripwire-versions-list
The other two commands do not work in our instance, as a result currently werent developed.

## Screenshots
Tripwire incident in 6.0.0
image

Tripwire incident with correct layout in 5.0.0- still in progress

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CommonTypes/IncidentFields/incidentfield-Device_Id.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"name": "Device Id",
+	"neverSetAsRequired": false,
+	"ownerOnly": false,
+	"placeholder": "",
+	"required": false,
+	"script": "",
+	"selectValues": null,
+	"sla": 0,
+	"system": false,
+	"systemAssociatedTypes": null,
```
Comment (bakatzir) [inline Packs/CommonTypes/IncidentFields/incidentfield-Device_Id.json:28]:
```suggestion
	"systemAssociatedTypes": ["Tripwire File Change"],
```

File: Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
+{
+  "brandName": "Tripwire",
+  "custom": true,
+  "defaultIncidentType": "Tripwire File Change",
+  "id": "631dc1af-b67a-451c-8efd-a2bab980177dn",
```
Comment (bakatzir) [inline Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json:5]:
```suggestion
  "id": "Tripwire",
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire_description.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+## Tripwire
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire_description.md:1]: please add how to get the required creds or/and to set the necessary permissions.

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.yml [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+description: Tripwire is a file integrity management (FIM), FIM monitors files and folders on systems and is triggered when they have
+  changed.
+display: Tripwire
+name: Tripwire
+script:
+  commands:
+    - arguments:
+        - default: false
+          description: Versions IDs given comma seperated.
+          isArray: false
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.yml:62]:
```suggestion
          isArray: true
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.yml [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+          name: version_oids
+          required: false
+          secret: false
+        - default: false
+          description: Elements IDs of elements versions to fetch, comma seperated.
+          isArray: false
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.yml:68]: also true?
please fix for all args

File: Packs/Tripwire/pack_metadata.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    "name": "Tripwire",
+    "description": "Tripwire is a file integrity managment(FIM), FIM is used to track files and folders on different systems and monitors their changes.",
+    "support": "xsoar",
+    "currentVersion": "1.0.0",
+    "author": "Cortex XSOAR",
+    "url": "https://www.paloaltonetworks.com/cortex",
+    "email": "",
+    "created": "2020-11-15T07:34:09Z",
+    "categories": [
+        "Analytics & SIEM"
```
Comment (bakatzir) [inline Packs/Tripwire/pack_metadata.json:11]: are you sure? @Arsenikr WDYT?

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+
+
+def test_module(client: Client, args: dict) -> str:
+    """Tests API connectivity and authentication'
+
+    :type client: ``dict``
+    :param args: contains the test modules arguments.
+
+    :type client: ``Client``
+    :param client: Tripwire client to use
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:241]:
```suggestion
    :type client: ``Client``
    :param client: Tripwire client to use

    :type params: dict
    :param params: contains the test modules parameters.
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+    params['time_detected_range'] = f'{last_fetch},{time_now}'
+    fetch_filter = filter_versions(params)
+
+    return params, fetch_filter, last_fetch
+
+
+''' COMMAND FUNCTIONS '''
+
+
+def test_module(client: Client, args: dict) -> str:
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:234]: i would call it params
```suggestion
def test_module(client: Client, params: dict) -> str:
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+    if args.get('tags'):
+        filters += f"tag={args.get('tags')}&"
+    if args.get('limit'):
+        filters += f"pageLimit={args.get('limit')}&"
+    if args.get('start'):
+        filters += f"pageStart={args.get('start')}&"
+    return filters
+
+
+def prepare_fetch(params: dict, first_fetch: str):
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:205]: if used only from the `fetch_incidents` command, why not call it from within there?
i think it will also allow you to return just the necessary param and not all of them. WDYT?

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+            next_run (``Dict[str, str]``): Contains the datetime str that will be
+                    used in ``last_fetch`` on the next fetch.
+            incidents (``List[dict]``): List of incidents that will be created in XSOAR
+
+    :rtype: ``Tuple[Dict[str, int], List[dict]]``
+    """
+
+    incidents: List[Dict[str, Any]] = []
+    last_fetch = datetime.strptime(last_fetch, DATE_FORMAT)
+    last_fetched_ids = demisto.getLastRun().get('fetched_ids', [])
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:386]: please add an inline explanation why you added it.

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+            continue
+
+        incident = {
+            'name': incident_name,
+            'occurred': incident_created_time.strftime(DATE_FORMAT),
+            'rawJSON': json.dumps(alert),
+        }
+        incidents.append(incident)
+        last_fetch = incident_created_time
+        fetched_ids.extend([alert.get('id')])
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:408]:
```suggestion
        if incident_created_time < last_fetch:
            continue

        incident_name = alert.get('id')

        if incident_name in last_fetched_ids:
            continue
        last_fetch = incident_created_time

        incident = {
            'name': incident_name,
            'occurred': incident_created_time.strftime(DATE_FORMAT),
            'rawJSON': json.dumps(alert),
        }
        incidents.append(incident)
        fetched_ids.extend([alert.get('id')])
```

Comment (bakatzir) [reply Packs/Tripwire/Integrations/Tripwire/Tripwire.py:408]: just re ordering for better readability.

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+
+
+class Client(BaseClient):
+    """Client class to interact with the service API
+
+    This Client implements API calls, and does not contain any Demisto logic.
+    Should only do requests and return data.
+    It inherits from BaseClient defined in CommonServer Python.
+    Most calls use _http_request() that handles proxy, SSL verification, etc.
+    For this HelloWorld implementation, no special attributes defined
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:41]:
```suggestion
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+    alerts = client.get_versions(fetch_filter)
+    alerts = alerts[:int(max_results)]
+    fetched_ids = []
+    for alert in alerts:
+        incident_created_time = datetime.strptime(alert.get('timeDetected'), '%Y-%m-%dT%H:%M:%S.000Z')
+
+        if incident_created_time < last_fetch:
+            continue
+
+        incident_name = alert.get('id')
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:396]: let's try to think of a more descriptive incident name than only the ID. maybe something like:
`File <node_id> Changed, latest version ID is <version_id>`

**Review by bakatzir — COMMENTED** [review]
Very clean code, good job.
please consider adding to the HR a msg about the total amount of each listed element/node/version

File: Packs/CommonTypes/IncidentFields/incidentfield-Device_Id.json (hunk 1 above)
Comment (orhovy) [reply Packs/CommonTypes/IncidentFields/incidentfield-Device_Id.json:28]: It should be null for new incident fields, according to david, only for modified incident fields this key should be the list of associated types.

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.yml (hunk 4 above)
Comment (orhovy) [reply Packs/Tripwire/Integrations/Tripwire/Tripwire.yml:62]: is is not array in purpose, as this argument will be sent as a string to the api comma separated.
as a result I didnt think that there is a reason converting into array and then to string again,
So I would like to set it to string comma separated and send it directly to the api.

Comment (bakatzir) [reply Packs/Tripwire/Integrations/Tripwire/Tripwire.yml:62]: I see. most of our customers are associating it differently, let's make sure we make it clear in the desc.

File: Packs/Tripwire/Integrations/Tripwire/Tripwire_description.md (hunk 3 above)
Comment (orhovy) [reply Packs/Tripwire/Integrations/Tripwire/Tripwire_description.md:1]: You just need your username and password

**Comment by bakatzir** [comment]
this PR is pending the SDK release that will include this PR: https://github.com/demisto/demisto-sdk/pull/907
per the addition of a category.

File: Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+          "complex": null,
+          "simple": "md5"
+        },
+        "Policy Description": {
+          "complex": null,
+          "simple": "ruleName"
+        },
+        "Policy ID": {
+          "complex": null,
+          "simple": "ruleId"
```
Comment (bakatzir) [inline Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json:33]:
```suggestion
          "simple": "policyid"
```

File: Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+          "complex": null,
+          "simple": "elementName"
+        },
+        "MD5": {
+          "complex": null,
+          "simple": "md5"
+        },
+        "Policy Description": {
+          "complex": null,
+          "simple": "ruleName"
```
Comment (bakatzir) [inline Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json:29]:
```suggestion
          "simple": "policydescription"
```

File: Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  "defaultIncidentType": "Tripwire File Change",
+  "id": "Tripwire",
+  "isDefault": false,
+  "mapping": {
+    "Tripwire File Change": {
+      "dontMapEventToLabels": true,
+      "internalMapping": {
+        "Device Id": {
+          "complex": null,
+          "simple": "nodeId"
```
Comment (bakatzir) [inline Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json:13]:
```suggestion
          "simple": "deviceid"
```

File: Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        },
+        "Device Name": {
+          "complex": null,
+          "simple": "nodeName"
```
Comment (bakatzir) [inline Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json:17]:
```suggestion
          "simple": "devicename"
```

File: Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        },
+        "File Name": {
+          "complex": null,
+          "simple": "elementName"
```
Comment (bakatzir) [inline Packs/Tripwire/Classifiers/classifier-Tripwire_5_9_9.json:21]: please fix

File: Packs/Tripwire/pack_metadata.json [hunk 19]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Tripwire",
+    "description": "Tripwire is a file integrity managment(FIM), FIM is used to track files and folders on different systems and monitors their changes.",
```
Comment (bakatzir) [inline Packs/Tripwire/pack_metadata.json:3]:
```suggestion
    "description": "Tripwire is a file integrity managment(FIM),used to track files and folders on different systems and monitors their changes.",
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 20]
```diff
@@ -0,0 +1,510 @@
...
+        elif command == 'test-module':
+            return_results(test_module(client, demisto.params()))
+
+        elif command == 'fetch-incidents':
+            max_fetch = params.get('max_fetch', 10)
+            next_run, incidents = fetch_incidents(client=client, params=demisto.params(), max_results=max_fetch)
+            demisto.setLastRun(next_run)
+            demisto.incidents(incidents)
+        else:
+            raise NotImplementedError(f'{command} is not an existing Tripwire command')
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:503]:
```suggestion
        if command in commands:
            return_results(commands[command](client, demisto.args()))
        elif command == 'test-module':
            return_results(test_module(client, demisto.params()))
        elif command == 'fetch-incidents':
            max_fetch = params.get('max_fetch', 10)
            next_run, incidents = fetch_incidents(client=client, params=demisto.params(), max_results=max_fetch)
            demisto.setLastRun(next_run)
            demisto.incidents(incidents)
        else:
            raise NotImplementedError(f'{command} is not an existing Tripwire command')
```

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 21]
```diff
@@ -0,0 +1,510 @@
...
+            filters += f"nodeLabel={node_label}&"
+    if args.get('version_hashes'):
+        for hash in argToList(args.get('version_hashes')):
+            filters += f"hash={hash}&"
+    if args.get('baseline_version_ids'):
+        for baseline_version in argToList(args.get('baseline_version_ids')):
+            filters += f"baselineVersion={baseline_version}&"
+    if args.get('time_detected_range'):
+        start, end = argToList(args.get('time_detected_range'))
+        if not re.search(DATE_REG, start) and not re.search(DATE_REG, end):
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:136]: why did you add that? do not we check that in `parse_date_range`?

Comment (orhovy) [reply Packs/Tripwire/Integrations/Tripwire/Tripwire.py:136]: no , `parse_date_rang` expects a relative time only  and fails on regular date time such as 2020-11-29T21:00:00Z

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.yml [hunk 22]
```diff
@@ -0,0 +1,538 @@
...
+      name: time_detetcted_range
+      required: false
+      secret: false
+    - default: false
+      description: |-
+        Time received range of element versions to fetch. comma seperated.
+        The first argument should be the start of the range and the second is the range.
+        The format of the start and end can be either relative e.g. "2 days, 1 day" or date time "2020-11-24T17:07:27Z,2020-11-25T17:07:27Z"
+      isArray: true
+      name: time_received_range
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.yml:131]: i'm contemplating if it is better to separate it to 2 args. what happens if the give you 3 or more args? it is a bit crunchy.

**Comment by ShirleyDenkberg** [comment]
@bakatzir Doc review completed

**Comment by ShirleyDenkberg** [comment]
@orhovy Please remember to mark the credentials as required.

File: Packs/Tripwire/Integrations/Tripwire/Tripwire_description.md (hunk 3 above)
Comment (bakatzir) [reply Packs/Tripwire/Integrations/Tripwire/Tripwire_description.md:1]: ok. no special permissions needed?

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.py [hunk 23]
```diff
@@ -0,0 +1,509 @@
...
+
+    if args.get('time_received_range'):
+        start, end = argToList(args.get('time_received_range'))
+        if not re.search(DATE_REG, start) and not re.search(DATE_REG, end):
+            try:
+                start = parse_date_range(start, date_format=DATE_FORMAT)[0]
+                end = parse_date_range(end, date_format=DATE_FORMAT)[0]
+            except Exception:
+                raise DemistoException(f"Please insert time range in relative format e.g. '1 day', '2 days' or in "
+                                       f"date format {DATE_FORMAT}")
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.py:153]: if we are keeping this arg, please check that this arg holds exactly 2 values, else print an appropriate error.

File: Packs/Tripwire/Integrations/Tripwire/README.md [hunk 24]
```diff
@@ -0,0 +1,6682 @@
+Tripwire is a file integrity management (FIM), monitors files and folders on systems and is triggered when they have changed.
+This integration was integrated and tested with version 1 of Tripwire
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/README.md:2]: just `1`? not even 1.0.0?

File: Packs/Tripwire/Integrations/Tripwire/Tripwire.yml [hunk 25]
```diff
@@ -0,0 +1,554 @@
...
+      isArray: false
+      name: end_detected_time
+      required: false
+      secret: false
+    - default: false
+      description: |-
+        End recieved time of element versions to fetch.
+        The format can be either relative e.g. "2 days" or date time "2020-11-24T17:07:27Z"
+      isArray: false
+      name: end_received_time
```
Comment (bakatzir) [inline Packs/Tripwire/Integrations/Tripwire/Tripwire.yml:160]:
```suggestion
      description: |-
        End received time of element versions to fetch.
        The format can be either relative e.g. "2 days" or date time "2020-11-24T17:07:27Z"
      isArray: false
      name: end_received_time
```

**Review by bakatzir — APPROVED** [review]
just fix the typo and revise the version in the readme
